### PR TITLE
Migrate to null safety

### PR DESCRIPTION
**Pull request by lrhn**
Only migrated the non-deprecated code, as an extra incentive to get people off deprecated code.

Did not update the version number to 2.0.0-nullsafety.0 because pkg:test depends on null-safety<2.0.0.
It should be upgraded to depend on >=1.9.3 <3.0.0, so we can release a pkg:package_config 2.0.0 without the deprecated code.
Until then, we're stuck in package:test-recursive-dependency-hell.

### CODE REVIEW

**Comment by jakemac53** [comment]
> Did not update the version number to 2.0.0-nullsafety.0 because pkg:test depends on null-safety<2.0.0.
> It should be upgraded to depend on >=1.9.3 <3.0.0, so we can release a pkg:package_config 2.0.0 without the deprecated code.
> Until then, we're stuck in package:test-recursive-dependency-hell.

This cycle only exists in dev dependencies, I would use dependency overrides to force the version of package:test. This is what glob, args, etc are doing.

The test package should not depend on a version range of its deps that allows both migrated and unmigrated versions (nor should any package), that seems very risky.

File: lib/discovery.dart
Comment (jakemac53) [inline lib/discovery.dart]: We should do the breaking version and delete these libs.

File: lib/src/discovery.dart
Comment (jakemac53) [inline lib/src/discovery.dart]: I am surprised this is required though, it seems like it should support a default?

```suggestion
    void onError(Object error),
```

File: lib/src/discovery.dart
Comment (jakemac53) [inline lib/src/discovery.dart:58]: It is pretty weird to have a nullable required positional argument like this... should we consider changing the api?

It seems like the only required argument should be `location` really and the others should be optional named args.

File: lib/src/package_config.dart
Comment (jakemac53) [inline lib/src/package_config.dart]:
```suggestion
          Object? extraData}) =>
```

(throughout if you accept the suggestion)

File: lib/src/package_config_json.dart
Comment (jakemac53) [inline lib/src/package_config_json.dart]: This has me thinking that `onError` should have a return type of `Never`? That would likely be breaking... and it would mean the user is still required to throw _something_ from `onError`, which may or may not be desirable.

File: lib/src/package_config_json.dart
Comment (jakemac53) [inline lib/src/package_config_json.dart]: We have been going with `Map<String, Object?>` for all the other packages because it makes the potential error conditions more explicit via explicit casts instead of dynamic calls.

File: lib/discovery.dart
Comment (lrhn) [reply lib/discovery.dart]: Seems the analyzer is still *using* the deprecated API.

File: lib/src/discovery.dart
Comment (lrhn) [reply lib/src/discovery.dart]: This is an internal function, I ensure that the default is added at the API edge.

File: lib/src/package_config.dart
Comment (lrhn) [reply lib/src/package_config.dart]: LGTM.

File: lib/src/package_config_json.dart
Comment (lrhn) [reply lib/src/package_config_json.dart]: I'll remove every instance of `dynamic`. And good riddance :)

File: lib/discovery.dart
Comment (jakemac53) [reply lib/discovery.dart]: Is that because they _need_ to use it or just haven't migrated?

File: lib/src/package_config_json.dart
Comment (lrhn) [reply lib/src/package_config_json.dart]: The `onError` is deliberately not required to throw. That allows the front-end to accumulate multiple errors instead of just bailing out on the first one.

File: lib/src/discovery.dart
Comment (jakemac53) [reply lib/src/discovery.dart]: ack, but I would remove the hint now as it is a part of the type :)

**Comment by lrhn** [comment]
Used dependency overrides and made it a 2.0.0 version without deprecated APIs.

Still can't be tested because the analyzer depends on the removed APIs.

File: lib/discovery.dart
Comment (lrhn) [reply lib/discovery.dart]: They are reading `.packages` files directly, then using the legacy functionality in this package to parse the file into a map. That's not supported directly by the new API. You can read a `.packages` file into a `PackageConfig` object using
```dart
loadPackageConfig(file, preferNewest: false);
```
The plain map from name to package root isn't supported any more because it makes no sense for `package_config.json` files.

Sam is working on it now: https://github.com/dart-lang/sdk/issues/44116

**Review by jakemac53 — COMMENTED** [review]
This generally LGTM, but we should wait until it can be tested to land it :).

File: lib/src/discovery.dart
Comment (lrhn) [reply lib/src/discovery.dart]: Done. I think I assumed the migration tool would remove it (and avoid guessing `?` there).

**Review by natebosch — APPROVED** [review]

File: lib/src/discovery.dart
Comment (lrhn) [reply lib/src/discovery.dart:58]: It's internal API. 
I usually don't bother with optional parameters for internal API, it's not there to be ergonomic, just to be clear and precise.

**Comment by lrhn** [comment]
Tested, landed, will publish.

**Comment by pq** [comment]
Curious what the status on publishing here is?  I notice the build is broken but maybe this expected?

**Comment by natebosch** [comment]
This had been blocked from being pulled into the SDK because of analyzer usage of a deprecated API.

That was migrated to a `lib/src/` API which still isn't a good end state, but I don't know if that is a hard blocker for getting pulled into the SDK.

@lrhn - is this ready to get updated in the SDK? Is it still blocked by the analyzer `lib/src` import? https://github.com/dart-lang/sdk/issues/44116

What is the next step to push on here?

**Comment by lrhn** [comment]
I think it's fine to pull into the SDK as-is.

We'll fix the src/ dependency at some later point.

**Pull request merged**
